File: test/render.test.js
import test from 'node:test';
import assert from 'node:assert/strict';
import { execFile, spawn } from 'node:child_process';
import { mkdtemp, readFile, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { promisify } from 'node:util';
import { getAsset } from 'gfm-addons';
import { getGfmAssetUrl, renderMarkdownToHtml } from '../index.js';

const execFileAsync = promisify(execFile);
const cliPath = fileURLToPath(new URL('../cli.js', import.meta.url));

function runCliWithInput(args, input) {
  return new Promise((resolve, reject) => {
    const child = spawn(process.execPath, [cliPath, ...args], {
      stdio: ['pipe', 'pipe', 'pipe'],
    });
    let stdout = '';
    let stderr = '';

    child.stdout.setEncoding('utf8');
    child.stderr.setEncoding('utf8');
    child.stdout.on('data', (chunk) => {
      stdout += chunk;
    });
    child.stderr.on('data', (chunk) => {
      stderr += chunk;
    });
    child.on('error', reject);
    child.on('close', (code) => {
      resolve({ code, stdout, stderr });
    });
    child.stdin.end(input);
  });
}

test('renderMarkdownToHtml returns a complete HTML document with remote assets by default', () => {
  const html = renderMarkdownToHtml('# Hello', { title: 'Hello <World>' });

  assert.match(html, /^<!doctype html>/);
  assert.match(html, /<html>/);
  assert.match(html, /<title>Hello &lt;World&gt;<\/title>/);
  assert.match(html, /<meta property="og:title" content="Hello &lt;World&gt;">/);
  assert.match(html, /<meta name="twitter:title" content="Hello &lt;World&gt;">/);
  assert.match(html, /<article class="markdown-body">/);
  assert.match(html, new RegExp(getAsset('ravel_gfm_css').remoteUrl.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')));
  assert.doesNotMatch(html, /gfm-addons\.js/);
  assert.doesNotMatch(html, /highlight-light\.css/);
});

test('renderMarkdownToHtml uses YAML front matter for head metadata without rendering it in the article', () => {
  const html = renderMarkdownToHtml(`---
title: YAML Title
description: YAML Description
canonical: https://example.test/post
cover: https://example.test/cover.png
date: 2026-06-10
update: 2026-06-11T10:20:30Z
---
# Visible`);
  const articleHtml = html.slice(html.indexOf('<article class="markdown-body">'), html.indexOf('</article>'));

  assert.match(html, /<title>YAML Title<\/title>/);
  assert.match(html, /<link rel="canonical" href="https:\/\/example\.test\/post">/);
  assert.match(html, /<meta name="description" content="YAML Description">/);
  assert.match(html, /<meta property="og:url" content="https:\/\/example\.test\/post">/);
  assert.match(html, /<meta property="og:image" content="https:\/\/example\.test\/cover\.png">/);
  assert.match(html, /<meta property="article:published_time" content="2026-06-10">/);
  assert.match(html, /<meta property="article:modified_time" content="2026-06-11T10:20:30Z">/);
  assert.match(html, /<meta name="twitter:card" content="summary_large_image">/);
  assert.match(html, /<meta name="twitter:image" content="https:\/\/example\.test\/cover\.png">/);
  assert.match(html, /Visible/);
  assert.doesNotMatch(articleHtml, /YAML Title/);
  assert.doesNotMatch(articleHtml, /YAML Description/);
  assert.doesNotMatch(articleHtml, /canonical:/);
});

test('renderMarkdownToHtml resolves title priority from options, YAML, heading, then empty', () => {
  const optionTitleHtml = renderMarkdownToHtml('---\ntitle: YAML Title\n---\n# Heading Title', { title: 'Option Title' });
  const yamlTitleHtml = renderMarkdownToHtml('---\ntitle: YAML Title\n---\n# Heading Title');
  const headingTitleHtml = renderMarkdownToHtml('# Heading Title');
  const emptyTitleHtml = renderMarkdownToHtml('');

  assert.match(optionTitleHtml, /<title>Option Title<\/title>/);
  assert.match(yamlTitleHtml, /<title>YAML Title<\/title>/);
  assert.match(headingTitleHtml, /<title>Heading Title<\/title>/);
  assert.match(emptyTitleHtml, /<title><\/title>/);
});

test('renderMarkdownToHtml resolves canonical priority and omits URL tags when absent', () => {
  const html = renderMarkdownToHtml('---\ncanonical: https://yaml.example/post\n---\n# Hello', {
    canonical: 'https://option.example/post',
  });
  const withoutCanonical = renderMarkdownToHtml('# Hello');

  assert.match(html, /<link rel="canonical" href="https:\/\/option\.example\/post">/);
  assert.match(html, /<meta property="og:url" content="https:\/\/option\.example\/post">/);
  assert.doesNotMatch(html, /https:\/\/yaml\.example\/post/);
  assert.doesNotMatch(withoutCanonical, /rel="canonical"/);
  assert.doesNotMatch(withoutCanonical, /property="og:url"/);
});

test('renderMarkdownToHtml resolves image priority and ignores relative Markdown images', () => {
  const coverHtml = renderMarkdownToHtml(`---
cover: https://example.test/cover.png
image: https://example.test/image.png
---
# Hello

![Body](https://example.test/body.png)`);
  const markdownImageHtml = renderMarkdownToHtml('# Hello\n\n![Body](https://example.test/body.png)');
  const relativeImageHtml = renderMarkdownToHtml('# Hello\n\n![Body](/body.png)');

  assert.match(coverHtml, /<meta property="og:image" content="https:\/\/example\.test\/cover\.png">/);
  assert.doesNotMatch(coverHtml, /https:\/\/example\.test\/image\.png/);
  assert.match(markdownImageHtml, /<meta property="og:image" content="https:\/\/example\.test\/body\.png">/);
  assert.doesNotMatch(relativeImageHtml, /og:image/);
  assert.match(relativeImageHtml, /<meta name="twitter:card" content="summary">/);
});

test('renderMarkdownToHtml uses a stable grayscale Picsum fallback image only when enabled', () => {
  const markdown = '# Hello\n\nBody without an image.';
  const defaultHtml = renderMarkdownToHtml(markdown);
  const fallbackHtml = renderMarkdownToHtml(markdown, { fallbackImage: true });
  const repeatedFallbackHtml = renderMarkdownToHtml(markdown, { fallbackImage: true });
  const yamlImageHtml = renderMarkdownToHtml('---\nimage: https://example.test/image.png\n---\n# Hello', {
    fallbackImage: true,
  });
  const imageMatch = fallbackHtml.match(/<meta property="og:image" content="(https:\/\/picsum\.photos\/seed\/[a-f0-9]{16}\/1200\/630\.jpg\?grayscale)">/);
  const repeatedImageMatch = repeatedFallbackHtml.match(/<meta property="og:image" content="(https:\/\/picsum\.photos\/seed\/[a-f0-9]{16}\/1200\/630\.jpg\?grayscale)">/);

  assert.doesNotMatch(defaultHtml, /picsum\.photos/);
  assert.ok(imageMatch);
  assert.ok(repeatedImageMatch);
  assert.equal(imageMatch[1], repeatedImageMatch[1]);
  assert.match(fallbackHtml, /<meta name="twitter:card" content="summary_large_image">/);
  assert.match(fallbackHtml, new RegExp(`<meta name="twitter:image" content="${imageMatch[1].replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}">`));
  assert.match(yamlImageHtml, /<meta property="og:image" content="https:\/\/example\.test\/image\.png">/);
  assert.doesNotMatch(yamlImageHtml, /picsum\.photos/);
});

test('renderMarkdownToHtml falls back to a 160 character plain text description', () => {
  const bodyText = `${'A'.repeat(100)} ${'B'.repeat(100)}`;
  const html = renderMarkdownToHtml(`# Title

\`\`\`js
console.log('skip me');
\`\`\`

${bodyText}`);
  const match = html.match(/<meta name="description" content="([^"]+)">/);

  assert.ok(match);
  assert.equal(match[1].length, 160);
  assert.equal(match[1], bodyText.slice(0, 160));
  assert.doesNotMatch(match[1], /skip me/);
});

test('renderMarkdownToHtml injects local TOC assets when enough headings exist', () => {
  const html = renderMarkdownToHtml('# One\n\n## Two', { assetMode: 'local' });

  assert.match(html, /href="\/asset\/gfm_addons_css"/);
  assert.match(html, /src="\/asset\/gfm_addons_js"/);
});

test('renderMarkdownToHtml injects highlight assets only for code blocks', () => {
  const html = renderMarkdownToHtml('```js\nconsole.log("hi")\n```', { assetMode: 'local' });

  assert.match(html, /href="\/asset\/highlight_light_css" media="\(prefers-color-scheme: light\)"/);
  assert.match(html, /href="\/asset\/highlight_dark_css" media="\(prefers-color-scheme: dark\)"/);
});

test('renderMarkdownToHtml supports slots, extra CSS, and body classes', () => {
  const html = renderMarkdownToHtml('# Hello', {
    extraCss: '.custom { color: red; }',
    bodyClass: 'custom-body',
    slots: {
      headEnd: '<meta name="x-test" content="1">',
      bodyStart: '<!-- raw hint -->',
      articleBefore: '<nav>before</nav>',
      articleAfter: () => '<footer>after</footer>',
      bodyEnd: '<script>window.done = true;</script>',
    },
  });

  assert.match(html, /<meta name="x-test" content="1">/);
  assert.match(html, /<body class="custom-body">/);
  assert.match(html, /<!-- raw hint -->/);
  assert.match(html, /<nav>before<\/nav>/);
  assert.match(html, /<footer>after<\/footer>/);
  assert.match(html, /\.custom \{ color: red; \}/);
});

test('renderMarkdownToHtml injects a sticky markdown footer', () => {
  const footerHtml = 'footer-e8c3a91f <a href="https://example.test/link-42">link-17b92</a>';
  const html = renderMarkdownToHtml('# One\n\n## Two', { footerHtml });
  const articleEndIndex = html.indexOf('</article>');
  const tocScriptIndex = html.indexOf('gfm-addons.js');
  const footerIndex = html.indexOf('<footer class="markdown-body post-footer">');

  assert.notEqual(footerIndex, -1);
  assert.ok(articleEndIndex < footerIndex);
  assert.ok(tocScriptIndex < footerIndex);
  assert.match(html, /min-height: 100vh;/);
  assert.match(html, /display: flex;/);
  assert.match(html, /flex-direction: column;/);
  assert.match(html, /\.post-footer \{\n    flex-shrink: 0;\n    margin-top: auto;\n    padding-top: 48px;\n    text-align: center;\n    font-size: 12px;\n  \}/);
  assert.match(html, /<footer class="markdown-body post-footer">\nfooter-e8c3a91f <a href="https:\/\/example\.test\/link-42">link-17b92<\/a>\n<\/footer>/);
});

test('renderMarkdownToHtml omits the sticky footer for blank footer HTML', () => {
  const html = renderMarkdownToHtml('# Hello', { footerHtml: '   ' });

  assert.doesNotMatch(html, /post-footer/);
  assert.doesNotMatch(html, /min-height: 100vh;/);
  assert.doesNotMatch(html, /display: flex;/);
});

test('getGfmAssetUrl supports custom resolver and local base URL', () => {
  assert.equal(
    getGfmAssetUrl('ravel_gfm_css', { assetMode: 'local', assetBaseUrl: '/static/assets' }),
    '/static/assets/ravel_gfm_css',
  );
  assert.equal(
    getGfmAssetUrl('ravel_gfm_css', { resolveAssetUrl: (asset) => `/custom/${asset.key}` }),
    '/custom/ravel_gfm_css',
  );
});

test('getGfmAssetUrl rejects unsupported asset modes', () => {
  assert.throws(
    () => getGfmAssetUrl('ravel_gfm_css', { assetMode: 'inline' }),
    /Unsupported assetMode: inline/,
  );
});

test('CLI prints help with --help and -h', async () => {
  const help = await execFileAsync(process.execPath, [cliPath, '--help']);
  const shortHelp = await execFileAsync(process.execPath, [cliPath, '-h']);

  assert.match(help.stdout, /^Usage: gfm-it \[file\] \[options\]/);
  assert.match(help.stdout, /--canonical <url>/);
  assert.match(help.stdout, /--fallback-image <true\|false>/);
  assert.match(shortHelp.stdout, /^Usage: gfm-it \[file\] \[options\]/);
});

test('CLI reads stdin when file is omitted', async () => {
  const result = await runCliWithInput(['--title', 'From stdin'], '# From stdin');

  assert.equal(result.code, 0);
  assert.match(result.stdout, /<title>From stdin<\/title>/);
  assert.match(result.stdout, /From stdin/);
  assert.equal(result.stderr, '');
});

test('CLI renders a file to an output path', async () => {
  const directory = await mkdtemp(join(tmpdir(), 'gfm-it-'));
  const input = join(directory, 'input.md');
  const output = join(directory, 'nested', 'output.html');
  await writeFile(input, '# File input');

  await execFileAsync(process.execPath, [cliPath, input, '--output', output, '--asset-mode', 'local']);
  const html = await readFile(output, 'utf8');

  assert.match(html, /File input/);
  assert.match(html, /href="\/asset\/ravel_gfm_css"/);
});

test('CLI accepts raw footer HTML', async () => {
  const result = await runCliWithInput(['--footer-html', '<strong>CLI footer</strong>'], '# From stdin');

  assert.equal(result.code, 0);
  assert.match(result.stdout, /<footer class="markdown-body post-footer">\n<strong>CLI footer<\/strong>\n<\/footer>/);
  assert.match(result.stdout, /min-height: 100vh;/);
  assert.equal(result.stderr, '');
});

test('CLI accepts a canonical URL', async () => {
  const result = await runCliWithInput(['--canonical', 'https://example.test/stdin'], '# From stdin');

  assert.equal(result.code, 0);
  assert.match(result.stdout, /<link rel="canonical" href="https:\/\/example\.test\/stdin">/);
  assert.match(result.stdout, /<meta property="og:url" content="https:\/\/example\.test\/stdin">/);
  assert.equal(result.stderr, '');
});

test('CLI accepts the fallback image boolean option', async () => {
  const result = await runCliWithInput(['--fallback-image', 'true'], '# From stdin');
  const falseResult = await runCliWithInput(['--fallback-image', 'false'], '# From stdin');

  assert.equal(result.code, 0);
  assert.match(result.stdout, /<meta property="og:image" content="https:\/\/picsum\.photos\/seed\/[a-f0-9]{16}\/1200\/630\.jpg\?grayscale">/);
  assert.match(result.stdout, /<meta name="twitter:card" content="summary_large_image">/);
  assert.equal(result.stderr, '');
  assert.equal(falseResult.code, 0);
  assert.doesNotMatch(falseResult.stdout, /picsum\.photos/);
  assert.equal(falseResult.stderr, '');
});

test('CLI rejects invalid fallback image boolean values', async () => {
  const result = await runCliWithInput(['--fallback-image', 'yes'], '# From stdin');

  assert.equal(result.code, 1);
  assert.match(result.stderr, /--fallback-image must be true or false, got: yes/);
});

File: render_test.go
package gfmit

import (
	"regexp"
	"strings"
	"testing"
)

func TestRenderMarkdownToHTMLReturnsCompleteDocumentWithRemoteAssets(t *testing.T) {
	html, err := RenderMarkdownToHTML("# Hello", RenderOptions{Title: "Hello <World>"})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	if !strings.HasPrefix(html, "<!doctype html>") {
		t.Fatalf("expected doctype, got %q", html)
	}
	assertContains(t, html, "<html>")
	assertContains(t, html, "<title>Hello &lt;World&gt;</title>")
	assertContains(t, html, `<meta property="og:title" content="Hello &lt;World&gt;">`)
	assertContains(t, html, MustAsset("ravel_gfm_css").RemoteURL)
	assertNotContains(t, html, "gfm-addons.js")
	assertNotContains(t, html, "highlight-light.css")
}

func TestRenderMarkdownToHTMLUsesYAMLMetadataWithoutRenderingFrontMatter(t *testing.T) {
	html, err := RenderMarkdownToHTML(`---
title: YAML Title
description: YAML Description
canonical: https://example.test/post
cover: https://example.test/cover.png
date: 2026-06-10
update: 2026-06-11T10:20:30Z
---
# Visible`, RenderOptions{})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	article := html[strings.Index(html, `<article class="markdown-body">`):strings.Index(html, "</article>")]
	assertContains(t, html, "<title>YAML Title</title>")
	assertContains(t, html, `<link rel="canonical" href="https://example.test/post">`)
	assertContains(t, html, `<meta name="description" content="YAML Description">`)
	assertContains(t, html, `<meta property="og:url" content="https://example.test/post">`)
	assertContains(t, html, `<meta property="og:image" content="https://example.test/cover.png">`)
	assertContains(t, html, `<meta property="article:published_time" content="2026-06-10">`)
	assertContains(t, html, `<meta property="article:modified_time" content="2026-06-11T10:20:30Z">`)
	assertContains(t, article, "Visible")
	assertNotContains(t, article, "YAML Title")
	assertNotContains(t, article, "canonical:")
}

func TestRenderMarkdownToHTMLSupportsLocalAssetsAndDynamicResources(t *testing.T) {
	html, err := RenderMarkdownToHTML("# One\n\n## Two\n\n```go\nfmt.Println(\"hi\")\n```\n\n$$\na+b\n$$", RenderOptions{AssetMode: "local"})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertContains(t, html, `href="/asset/ravel_gfm_css"`)
	assertContains(t, html, `href="/asset/gfm_addons_css"`)
	assertContains(t, html, `src="/asset/gfm_addons_js"`)
	assertContains(t, html, `href="/asset/highlight_light_css" media="(prefers-color-scheme: light)"`)
	assertContains(t, html, `href="/asset/highlight_dark_css" media="(prefers-color-scheme: dark)"`)
	assertContains(t, html, `src="/asset/highlight_js" defer`)
	assertContains(t, html, "katex.min.css")
}

func TestRenderMarkdownToHTMLSupportsInlineAssets(t *testing.T) {
	html, err := RenderMarkdownToHTML("# One\n\n## Two\n\n```go\nfmt.Println(\"hi\")\n```", RenderOptions{AssetMode: "inline"})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertContains(t, html, `<style data-gfm-asset="ravel_gfm_css">`)
	assertContains(t, html, `<style data-gfm-asset="gfm_addons_css">`)
	assertContains(t, html, `<script data-gfm-asset="gfm_addons_js">`)
	assertContains(t, html, `<style data-gfm-asset="highlight_light_css" media="(prefers-color-scheme: light)">`)
	assertContains(t, html, `<style data-gfm-asset="highlight_dark_css" media="(prefers-color-scheme: dark)">`)
	assertContains(t, html, `<script data-gfm-asset="highlight_js">`)
	assertNotContains(t, html, "cdn.jsdelivr.net")
	assertNotContains(t, html, `href="/asset/`)
	assertNotContains(t, html, `src="/asset/`)
}

func TestRenderMarkdownToHTMLNormalizesCSSAlias(t *testing.T) {
	html, err := RenderMarkdownToHTML("# GitHub", RenderOptions{AssetMode: "inline", CSS: "github"})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertContains(t, html, `<style data-gfm-asset="github_gfm_css">`)
}

func TestRenderMarkdownToHTMLAcceptsCSSHref(t *testing.T) {
	html, err := RenderMarkdownToHTML("# Remote", RenderOptions{CSS: "http://abc.com/a.css"})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertContains(t, html, `<link rel="stylesheet" href="http://abc.com/a.css">`)
	assertNotContains(t, html, "ravel-gfm.min.css")

	localHTML, err := RenderMarkdownToHTML("# Local", RenderOptions{CSS: "../css"})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertContains(t, localHTML, `<link rel="stylesheet" href="../css">`)

	queryHTML, err := RenderMarkdownToHTML("# Query", RenderOptions{CSS: "https://example.com/hi.css?raw=true"})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertContains(t, queryHTML, `<link rel="stylesheet" href="https://example.com/hi.css?raw=true">`)
}

func TestRenderMarkdownToHTMLUsesYAMLCSSOverride(t *testing.T) {
	aliasHTML, err := RenderMarkdownToHTML("---\ngfm_css: github\n---\n# YAML CSS", RenderOptions{AssetMode: "inline", CSS: "folio"})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertContains(t, aliasHTML, `<style data-gfm-asset="github_gfm_css">`)
	assertNotContains(t, aliasHTML, `data-gfm-asset="folio_gfm_css"`)

	hrefHTML, err := RenderMarkdownToHTML("---\ngfm_css: ../css\n---\n# YAML CSS", RenderOptions{AssetMode: "inline", CSS: "github"})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertContains(t, hrefHTML, `<link rel="stylesheet" href="../css">`)
	assertNotContains(t, hrefHTML, `data-gfm-asset="github_gfm_css"`)

	fallbackHTML, err := RenderMarkdownToHTML("---\ngfm_css: https://example.com/theme.js\n---\n# YAML CSS", RenderOptions{AssetMode: "inline", CSS: "github"})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertContains(t, fallbackHTML, `<style data-gfm-asset="github_gfm_css">`)
	assertNotContains(t, fallbackHTML, "theme.js")
}

func TestRenderMarkdownToHTMLKeepsYAMLThemeAssetsModeAware(t *testing.T) {
	localThemeHTML, err := RenderMarkdownToHTML("---\ngfm_css: github\n---\n# YAML CSS", RenderOptions{AssetMode: "local", CSS: "folio"})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertContains(t, localThemeHTML, `<link rel="stylesheet" href="/asset/github_gfm_css">`)
	assertNotContains(t, localThemeHTML, "folio_gfm_css")

	localHrefHTML, err := RenderMarkdownToHTML("---\ngfm_css: ../css\n---\n# YAML CSS", RenderOptions{AssetMode: "local", CSS: "github"})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertContains(t, localHrefHTML, `<link rel="stylesheet" href="../css">`)
	assertNotContains(t, localHrefHTML, `/asset/github_gfm_css`)
}

func TestRenderMarkdownToHTMLSupportsSlotsExtraCSSBodyClassAndFooter(t *testing.T) {
	html, err := RenderMarkdownToHTML("# One\n\n## Two", RenderOptions{
		AssetMode:  "local",
		ExtraCSS:   ".custom { color: red; }",
		BodyClass:  "custom-body",
		FooterHTML: " footer-e8c3a91f <a href=\"https://example.test/link-42\">link-17b92</a> ",
		Slots: RenderSlots{
			HeadEnd:       `<meta name="x-test" content="1">`,
			BodyStart:     "<!-- raw hint -->",
			ArticleBefore: "<nav>before</nav>",
			ArticleAfter:  "<footer>after</footer>",
			BodyEnd:       "<script>window.done = true;</script>",
		},
	})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertContains(t, html, `<meta name="x-test" content="1">`)
	assertContains(t, html, `<body class="custom-body">`)
	assertContains(t, html, "<!-- raw hint -->")
	assertContains(t, html, "<nav>before</nav>")
	assertContains(t, html, "<footer>after</footer>")
	assertContains(t, html, ".custom { color: red; }")
	assertContains(t, html, `<footer class="markdown-body post-footer">
footer-e8c3a91f <a href="https://example.test/link-42">link-17b92</a>
</footer>`)
	assertContains(t, html, "min-height: 100vh;")
	assertContains(t, html, "display: flex;")
	assertContains(t, html, "flex-direction: column;")
	articleEndIndex := strings.Index(html, "</article>")
	tocScriptIndex := strings.Index(html, "/asset/gfm_addons_js")
	footerIndex := strings.Index(html, `<footer class="markdown-body post-footer">`)
	if articleEndIndex == -1 || tocScriptIndex == -1 || footerIndex == -1 || !(articleEndIndex < tocScriptIndex && tocScriptIndex < footerIndex) {
		t.Fatalf("expected footer after article and dynamic scripts, got %q", html)
	}
}

func TestRenderMarkdownToHTMLOmitsBlankFooter(t *testing.T) {
	html, err := RenderMarkdownToHTML("# Hello", RenderOptions{FooterHTML: "   "})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertNotContains(t, html, "post-footer")
	assertNotContains(t, html, "min-height: 100vh;")
}

func TestRenderMarkdownToHTMLPreservesRawHTMLAndUnclosedFrontMatter(t *testing.T) {
	html, err := RenderMarkdownToHTML("Hello\n\n<script>alert('xss')</script>", RenderOptions{})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertContains(t, html, "<script>alert('xss')</script>")

	unclosed, err := RenderMarkdownToHTML("---\nnot: closed\nbody", RenderOptions{})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertContains(t, unclosed, "<hr>")
	assertContains(t, unclosed, "not: closed")
}

func TestRenderMarkdownToHTMLEnablesFallbackSocialImage(t *testing.T) {
	html, err := RenderMarkdownToHTML("# Hello", RenderOptions{FallbackImage: true})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	pattern := regexp.MustCompile(`<meta property=\"og:image\" content=\"https://picsum\.photos/seed/[a-f0-9]{16}/1200/630\.jpg\?grayscale\">`)
	if !pattern.MatchString(html) {
		t.Fatalf("expected fallback og:image, got %q", html)
	}
	assertContains(t, html, `<meta name="twitter:card" content="summary_large_image">`)
}

func TestRenderMarkdownToHTMLUsesCustomAssetResolver(t *testing.T) {
	html, err := RenderMarkdownToHTML("# Hello", RenderOptions{
		ResolveAssetURL: func(asset Asset) (string, error) {
			return "/custom/" + asset.Key, nil
		},
	})
	if err != nil {
		t.Fatalf("RenderMarkdownToHTML() error = %v", err)
	}
	assertContains(t, html, `href="/custom/ravel_gfm_css"`)

	_, err = RenderMarkdownToHTML("# Hello", RenderOptions{
		ResolveAssetURL: func(asset Asset) (string, error) { return "", nil },
	})
	if err == nil || !strings.Contains(err.Error(), "ResolveAssetURL must return a non-empty string") {
		t.Fatalf("expected empty resolver error, got %v", err)
	}
}

func TestRenderMarkdownToHTMLRejectsUnsupportedAssetMode(t *testing.T) {
	_, err := RenderMarkdownToHTML("# Hello", RenderOptions{AssetMode: "bad"})
	if err == nil || !strings.Contains(err.Error(), "unsupported asset mode") {
		t.Fatalf("expected unsupported asset mode error, got %v", err)
	}
}

func TestRenderMarkdownToHTMLRejectsUnsupportedCSSAsset(t *testing.T) {
	for _, css := range []string{"missing", "highlight_light_css"} {
		_, err := RenderMarkdownToHTML("# Hello", RenderOptions{CSS: css})
		if err == nil || !strings.Contains(err.Error(), "unsupported CSS asset: "+css) {
			t.Fatalf("expected unsupported CSS asset error for %q, got %v", css, err)
		}
	}
	_, err := RenderMarkdownToHTML("# Hello", RenderOptions{AssetMode: "inline", CSS: "http://abc.com/a.css"})
	if err == nil || !strings.Contains(err.Error(), "CSS hrefs require asset mode remote") {
		t.Fatalf("expected CSS href asset mode error, got %v", err)
	}
	_, err = RenderMarkdownToHTML("# Hello", RenderOptions{AssetMode: "remote", CSS: "https://cdn.example.com/theme.js"})
	if err == nil || !strings.Contains(err.Error(), "unsupported CSS URL") {
		t.Fatalf("expected unsupported CSS URL error, got %v", err)
	}
}

func assertContains(t *testing.T, haystack, needle string) {
	t.Helper()
	if !strings.Contains(haystack, needle) {
		t.Fatalf("expected %q to contain %q", haystack, needle)
	}
}

func assertNotContains(t *testing.T, haystack, needle string) {
	t.Helper()
	if strings.Contains(haystack, needle) {
		t.Fatalf("expected %q not to contain %q", haystack, needle)
	}
}
